### PR TITLE
`Development`: Extract plagiarism detection service

### PR DESCRIPTION
**Pull request by jakubriegel**
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://ls1intum.github.io/Artemis/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://ls1intum.github.io/Artemis/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.
#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context
Plagiarism detection logic is distributed in controllers for each exercise type resources. This logic has a common domain and could be stored in a dedicated class. The current state will also require a copy&paste approach in case on implementing new triggers for plagiarism detection, like continuous plagiarism control.

### Description
This PR introduces a common `PlagiarismDetectionService` which own all plagiarism checks related logic. This class is implemented using the code moved from `ModelingExerciseResource`, `TextExerciseResource` and `ProgrammingExercisePlagiarismResource` classes. 

There is a new `PlagarismDetectionConfig` class, which holds the configuration for plagiarism checks. This is the first step that enables this config to be stored on the server, which will be necessary for the continuous plagiarism control in the future.

### Steps for Testing
Prerequisites:
- 1 Instructor and 2 Students
- 1 programming, 1 modeling and 1 text exercise

1. Create very similar submission to each of 3 exercises from both user accounts.
2. Go to exercise manage page and click on _Check plagiarism_.
3. Run plagiarism detection wich default settings.
4. Verify that plagiarism results are shown.
5. Run plagiarism detection wich similarity threshold changed to 100%.
6. Verify that **no** plagiarism result is shown.

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| PlagiarismDetectionService | 100% | ✅                           |

### CODE REVIEW

**Review by pal03377 — APPROVED** [review]
Tested on TS2 in the testing session. We found some weird case where the similarity of two programming exercises that are the same was only ~70%, but we believe this to be no issue of this PR.

**Review by Kroko-fant — APPROVED** [review]
Tested in the testing-session. We found bugs not related to the pr and Jakub created an issue #7174

**Review by Kroko-fant — APPROVED** [review]

**Review by JohannesStoehr — APPROVED** [review]
Refactoring looks good to me

**Review by MarkusPaulsen — DISMISSED** [review]
Code looks good + Tested on TS1

File: src/main/java/de/tum/in/www1/artemis/domain/plagiarism/PlagiarismDetectionConfig.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
+package de.tum.in.www1.artemis.domain.plagiarism;
+
+/**
+ * Stores configuration for plagiarism detection.
+ * This class is not a record as in the future (when cpc will be implemented) it will be extended to a DomainObject and represent a database entry.
+ */
+public class PlagiarismDetectionConfig {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/domain/plagiarism/PlagiarismDetectionConfig.java:7]: please use a record for this class

**Pull request merged**